src/io.ts

```diff
@@ -0,0 +1,47 @@
+import { readFile, stat, unlink } from "fs/promises";
+import { STALE_TIMEOUT } from "./constants";
+import { getRequestPath, getResponsePath } from "./paths";
+import { Request, Response } from "./types";
+import { writeJSONExclusive } from "./fileUtils";
+
+/**
+ * Reads the JSON-encoded request from the request file, unlinking the file
+ * after reading.
+ * @returns A promise that resolves to a Response object
+ */
+export async function readRequest(): Promise<Request> {
+  const requestPath = getRequestPath();
+  const request = JSON.parse(await readFile(requestPath, "utf-8"));
+  await unlink(requestPath);
+
+  return request;
+}
+
+/**
+ * Writes the response to the response file as JSON.  If a stale response file
+ * exists, it will be removed.  If a non-stale response file exists, assumes
+ * there is another VSCode instance mysteriously trying to handle the request
+ * as well and fails.
+ * @param response The response object to JSON-encode and write to disk
+ */
+export async function writeResponse(response: Response) {
+  const responsePath = getResponsePath();
+
+  try {
+    await writeJSONExclusive(responsePath, response);
+  } catch (err) {
+    if (err.code !== "EEXIST") {
+      throw err;
+    }
+
+    const stats = await stat(responsePath);
```

Comment:
Will this stat throw an exception in the weird case where the file existed on line 31 but was deleted by now? If the file is already gone, you probably want to skip directly to the second `writeJsonExclusive`.

Similarly, will the `unlink` below fail if the file was already deleted? Probably don't care about errors there either, you can catch any major issues in the second writeJson

Reply:
Do we really want to succeed in that case? That would only ever happen in the case of a race, right?

Reply:
You're writing a response though, and the response is still doubly safe to write in this case because it has a UUID check, and exclusive open. The files might get deleted for some other reason, like the user opening a vscode window at a bad time.